### ISSUE TITLE
sympy.TableForm produces invalid math-mode latex

### ISSUE THREAD

**Issue by eric-wieser** [sympy/sympy#20063]
The contract of `LatexPrinter`'s `_latex` method is to return math-mode latex. However, for `TableForm` it returns in text mode, which leads to invalid latex for code like
```python
from sympy import latex, TableForm

t = TableForm([[1, 2]])
latex([t])
```
which gives
```latex
\left[ \begin{tabular}{l l}
$1$ & $2$ \\
\end{tabular}\right]
```

**Comment by sayandip18** [comment]
```
>>> from sympy import *
>>> t = TableForm([[1,2]])
>>> latex(t)
'\\begin{tabular}{l l}\n$1$ & $2$ \\\\\n\\end{tabular}'
>>> print(t.as_latex())
\begin{tabular}{l l}
$1$ & $2$ \\
\end{tabular}
>>> latex([t])
'\\left[ \\begin{tabular}{l l}\n$1$ & $2$ \\\\\n\\end{tabular}\\right]'
```

Apparently, `latex(t)` or `print(t.as_latex())` gives result different from `latex([t])`, because `[t]` is actually a `list`, not a `TableForm`

```
>>> type(t)
<class 'sympy.printing.tableform.TableForm'>
>>> type([t])
<class 'list'>
```

Also, what `latex(t)` or `print(t.as_latex())` returns is precisely what is written [here](https://github.com/sympy/sympy/blob/bdb49c4abfb35554a3c8ce761696ffff3bb837fe/sympy/printing/tableform.py#L25) in `sympy/tableform.py`. Probably, I am missing something, but appears to me that the function is working just fine for `TableForm`

**Comment by eric-wieser** [comment]
@sayandip18, the point is that `latex(...)` must always return something that is legal to put between `$`. You're right that `[t]` is a list, I was just using that as an example to show _why_ `_latex` must behave that way.

Here's a simpler bug that doesn't require `[]` to appear:
```python
>>> from sympy import *
>>> t = TableForm([[1,2]])
>>> print(latex(t, mode='equation'))
\begin{equation}\begin{tabular}{l l}
$1$ & $2$ \\   % illegal - $ can't appear in \begin{equation}
\end{tabular}\end{equation}
```

**Comment by sayandip18** [comment]
Oh, I get it now. I'll start a PR soon.

**Comment by eric-wieser** [comment]
I'm not sure if this is fixable - tabular isn't allowed inside math mode.

**Comment by eric-wieser** [comment]
Deprecating the `_latex` method and requiring users to use `as_latex()` might be reasonable

**Comment by sayandip18** [comment]
How about modifying `latex()` so that whenever the expression in argument is a `TableForm`, the `as_latex()` method is invoked?

Also, `_latex` method is the same as `latex()`, right?

**Comment by eric-wieser** [comment]
> Also, _latex method is the same as latex(), right?

Yes, `latex(tf)` calls `tf._latex`. The problem is the return value of `_latex` is illegal - it is like returning `-1`  from `__len__`.

**Comment by oscargus** [comment]
I'm not overly familiar with TableForm, but I think it is fair to say that this may be a case of garbage in, garbage out.

What about if you integrate a TableForm? Should it give valid LaTeX code? The whole purpose of TableForm seems to be to print tables.

One may of course argue if a list is a mathematical expression as such, but considering what SymPy is used for, I think it is fair to assume that it is.

**Comment by oscargus** [comment]
A third option is that `as_latex()` returns a `tabular` and `_latex()` returns an `array`. Should be quite easy by passing optional keyword arguments to `_latex`.

**Comment by eric-wieser** [comment]
> As option is that the latex-printer sets a flag to indicate if the top level object is a `TableForm` and then based on that decides on if it should be printed using `array` or `tabular`.

Here's how I think this should work:

* `TableForm._latex` always returns an `array` (this function must always return math-mode latex)
* `TableForm._repr_latex_` (the jupyter hook) returns the `tabular` (this function must always return text-mode latex)
* `sympy.latex` when `mode != 'plain'` should look at the top-level object, and use `_repr_latex_` instead of `_latex`

This last point might require some care so as not to interfere with the sympy Jupyter integration.

**Comment by ccatrett** [comment]
Hi, I am a first time contributor and I was wondering if this problem was still open and if there were any pointers for getting started on trying to fix this issue, i.e., what files to investigate, how to start testing, etc.

Any help would be appreciated. Thanks.

**Comment by ccatrett** [comment]
@eric-wieser I have done some digging into the code discussed. however, I was wondering if you could cite your claims about the API for `_latex` and `_repr_latex_`? I would like to investigate this more and I think those would be helpful resources.

**Comment by Mr-Neutr0n** [comment]
Opened a fix in #29152.

The root cause is that `_latex` returns `\begin{tabular}` which is a text-mode environment, but the contract of `_latex` is to return math-mode LaTeX. Switching to `\begin{array}` (the math-mode equivalent) and dropping the `$...$` cell wrapping resolves the invalid nesting when TableForm appears inside other expressions or when using `mode='equation'`/`mode='inline'`.
